librz/io/p/io_srec.c

```diff
@@ -0,0 +1,655 @@
+// SPDX-FileCopyrightText: 2021 Manolache183 <manolache.alexandru8@gmail.com>
+// SPDX-FileCopyrightText: 2021 swym  <0xfd000000@gmail.com>
+// SPDX-License-Identifier: LGPL-3.0-only
+
+/*
+*** SREC format description : every line follows this pattern
+	S / Type / Bytecount / Address / Data / Checksum
+
+	There are a bunch of types, from which one (S4) is reserved
+
+// source : https://en.wikipedia.org/wiki/SREC_(file_format)
+
+**** example records
+
+S00F000068656C6C6F202020202000003C
+S11F00007C0802A6900100049421FFF07C6C1B787C8C23783C6000003863000026
+S11F001C4BFFFFE5398000007D83637880010014382100107C0803A64E800020E9
+S5030003F9
+S9030000FC
+*/
+
+#include <rz_io.h>
+#include <rz_lib.h>
+
+#define NMAX             100
+#define SREC_PATH_PREFIX "srec://"
+
+// struct RzSREC : holds sparse buffer + its own fd, for internal management
+typedef struct {
+	st32 fd;
+	RzBuffer *rbuf;
+} RzSREC;
+
+static st32 write_record_S3(FILE *fd, ut32 start_addr, ut8 *b, ut8 size) {
+	ut8 recordSize = 4 + size + 1;
+
+	ut8 cks = recordSize;
+	cks += start_addr & 0xff;
+	cks += (start_addr >> 8) & 0xff;
+	cks += (start_addr >> 16) & 0xff;
+	cks += start_addr >> 24;
+
+	fprintf(fd, "S3%02x%08x", recordSize, start_addr);
+
+	for (int j = 0; j < size; j++) {
+		cks += b[j];
+		fprintf(fd, "%02x", b[j]);
+	}
+
+	fprintf(fd, "%02x\n", cks);
+
+	return 1;
+}
+
+static st32 __write(RzIO *io, RzIODesc *fd, const ut8 *buf, st32 count) {
+
+	const char *pathname;
+	FILE *out;
+	RzSREC *rih;
+	RzBufferSparse *rbs;
+	RzListIter *iter;
+
+	if (!fd || !fd->data || (fd->perm & RZ_PERM_W) == 0 || count <= 0) {
+		return -1;
+	}
+
+	rih = fd->data;
+	pathname = fd->name + 7;
+	out = rz_sys_fopen(pathname, "w");
+
+	// random starting record
+	fprintf(out, "S00F000068656C6C6F202020202000003C\n");
```

Comment:
why this is static?

Reply:
Didn't think much about this, ihex writing function is also static.
Please accept the bin test request so that the it builds.

Reply:
```
This record contains vendor specific ASCII text comment represented as a series of hex digit pairs. It is common to see the data for this record in the format of a null-terminated string. The text data can be anything including a mixture of the following information: file/module name, version/revision number, date/time, product name, vendor name, memory designator on PCB, copyright notice. It is common to see: 48 44 52 which is the ASCII H, D, and R - "HDR".[3] 
```
so the hex value says `hello` so i guess is ok